File: docs_crawler/crawler.py
from __future__ import annotations

import logging
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path
from typing import Dict, Iterable, List, Optional, Set

from .discovery import DiscoveredDoc, discover_documents
from .http_client import FetchResult, fetch_bytes
from .models import DocumentRef, Manifest, ManifestDocument, SourceConfig
from .storage import delete_document, load_manifest, save_manifest, write_document
from .util import (
    CrawlStats,
    collapse_index_path,
    ensure_within_dir,
    iso_now,
    normalize_url,
    safe_relative_path,
    sha256_bytes,
    split_language_from_path,
)

logger = logging.getLogger(__name__)


class Crawler:
    def __init__(
        self,
        *,
        allowed_extensions: Iterable[str],
        user_agent: str,
        timeout_seconds: int,
        max_workers: int,
    ) -> None:
        self.allowed_extensions = list(allowed_extensions)
        self.user_agent = user_agent
        self.timeout_seconds = timeout_seconds
        self.max_workers = max_workers

    def crawl_source(
        self,
        source: SourceConfig,
        *,
        prune: bool = False,
        dry_run: bool = False,
    ) -> CrawlStats:
        output_root = source.output_dir
        logger.info("crawl start source=%s output=%s", source.id, output_root)
        if dry_run:
            logger.info("dry run enabled source=%s", source.id)
        if prune:
            logger.info("prune enabled source=%s", source.id)
        allowed_extensions = (
            source.allowed_extensions
            if source.allowed_extensions is not None
            else self.allowed_extensions
        )
        manifest_path = output_root / ".manifest.json"
        manifest = load_manifest(manifest_path)
        if manifest is None:
            manifest = Manifest(
                version=1,
                source_id=source.id,
                generated_at=iso_now(),
                entrypoints=[ep.__dict__ for ep in source.entrypoints],
                documents={},
            )
        self._normalize_manifest_paths(manifest, output_root)

        discovered_urls: Set[str] = set()
        discovered_language_hints: Dict[str, Set[str]] = {}
        discovered_in_order: List[DiscoveredDoc] = []
        for entrypoint in source.entrypoints:
            logger.info("discover start source=%s type=%s", source.id, entrypoint.type)
            cache_dir: Optional[Path] = None
            if entrypoint.type == "github_tree":
                cache_dir = (
                    Path(entrypoint.git_cache_dir)
                    if entrypoint.git_cache_dir
                    else source.output_dir / ".git-cache"
                )
            discovered_batch = discover_documents(
                entrypoint,
                allowed_extensions=allowed_extensions,
                timeout_seconds=self.timeout_seconds,
                user_agent=self.user_agent,
                default_language=source.default_language,
                cache_dir=cache_dir,
            )
            before = len(discovered_urls)
            for discovered in discovered_batch:
                url = normalize_url(discovered.url)
                if url not in discovered_urls:
                    discovered_urls.add(url)
                    discovered_in_order.append(
                        DiscoveredDoc(
                            url=url,
                            language_hint=discovered.language_hint,
                            relative_path=discovered.relative_path,
                        )
                    )
                if discovered.language_hint:
                    discovered_language_hints.setdefault(url, set()).add(discovered.language_hint)
            added = len(discovered_urls) - before
            logger.info("discover done source=%s added=%d total=%d", source.id, added, len(discovered_urls))

        docs = self._build_document_refs(source, discovered_in_order, discovered_language_hints)
        stats = CrawlStats(discovered=len(docs))

        results: Dict[str, FetchResult] = {}
        total_docs = len(docs)
        logger.info("fetch start source=%s total=%d workers=%d", source.id, total_docs, self.max_workers)
        with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
            futures = {}
            for doc in docs:
                previous = manifest.documents.get(doc.url)
                if_none_match = None
                if_modified_since = None
                if previous is not None:
                    previous_path = output_root / previous.local_path if previous.local_path else None
                    if previous_path is not None:
                        previous_path = ensure_within_dir(output_root, previous_path)
                    target_path = ensure_within_dir(output_root, doc.local_path)
                    cache_ok = (previous_path is not None and previous_path.exists()) or target_path.exists()
                    if cache_ok:
                        if_none_match = previous.etag
                        if_modified_since = previous.last_modified

                futures[
                    executor.submit(
                        fetch_bytes,
                        doc.url,
                        timeout_seconds=self.timeout_seconds,
                        user_agent=self.user_agent,
                        if_none_match=if_none_match,
                        if_modified_since=if_modified_since,
                    )
                ] = doc

            completed = 0
            progress_every = self._progress_interval(total_docs)
            for future in as_completed(futures):
                doc = futures[future]
                try:
                    results[doc.url] = future.result()
                except Exception as exc:  # pragma: no cover
                    results[doc.url] = FetchResult(
                        url=doc.url, status=0, headers={}, body=None, error=f"Unhandled error: {exc}"
                    )
                completed += 1
                if completed % progress_every == 0 or completed == total_docs:
                    logger.info("fetch progress source=%s %d/%d", source.id, completed, total_docs)

        now = iso_now()
        for doc in docs:
            result = results.get(doc.url)
            if result is None:
                continue

            previous = manifest.documents.get(doc.url)
            if result.status == 304 and previous is not None:
                target_path = ensure_within_dir(output_root, doc.local_path)
                previous_path = output_root / previous.local_path if previous.local_path else None
                if previous_path is not None:
                    previous_path = ensure_within_dir(output_root, previous_path)

                if not target_path.exists() and previous_path is not None and previous_path.exists():
                    if not dry_run:
                        target_path.parent.mkdir(parents=True, exist_ok=True)
                        previous_path.replace(target_path)
                    previous.local_path = str(target_path.relative_to(output_root).as_posix())
                    previous.checked_at = now
                    previous.status = "moved"
                    stats = CrawlStats(
                        discovered=stats.discovered,
                        fetched=stats.fetched,
                        updated=stats.updated + 1,
                        unchanged=stats.unchanged,
                        errors=stats.errors,
                        pruned=stats.pruned,
                    )
                    continue

                if not target_path.exists():
                    refreshed = fetch_bytes(
                        doc.url,
                        timeout_seconds=self.timeout_seconds,
                        user_agent=self.user_agent,
                    )
                    result = refreshed
                else:
                    previous.checked_at = now
                    previous.status = "not_modified"
                    stats = CrawlStats(
                        discovered=stats.discovered,
                        fetched=stats.fetched,
                        updated=stats.updated,
                        unchanged=stats.unchanged + 1,
                        errors=stats.errors,
                        pruned=stats.pruned,
                    )
                    continue

                previous.checked_at = now
                previous.status = "refetched"

            if result.status != 200 or result.body is None:
                logger.warning(
                    "fetch error source=%s url=%s status=%s error=%s",
                    source.id,
                    doc.url,
                    result.status,
                    result.error or "",
                )
                entry = previous or ManifestDocument(
                    url=doc.url,
                    local_path=str(doc.local_path.relative_to(output_root).as_posix()),
                    language=doc.language,
                    sha256="",
                    bytes=0,
                )
                entry.checked_at = now
                entry.status = "error"
                entry.error = result.error or f"status={result.status}"
                manifest.documents[doc.url] = entry
                stats = CrawlStats(
                    discovered=stats.discovered,
                    fetched=stats.fetched,
                    updated=stats.updated,
                    unchanged=stats.unchanged,
                    errors=stats.errors + 1,
                    pruned=stats.pruned,
                )
                continue

            body = result.body
            sha = sha256_bytes(body)
            local_path = ensure_within_dir(output_root, doc.local_path)
            relative_local_path = str(local_path.relative_to(output_root).as_posix())

            path_changed = previous is not None and bool(previous.local_path) and previous.local_path != relative_local_path
            changed = path_changed or previous is None or previous.sha256 != sha or not local_path.exists()
            if changed and not dry_run:
                if path_changed and previous is not None and previous.local_path:
                    previous_path = ensure_within_dir(output_root, output_root / previous.local_path)
                    if previous_path.exists() and previous_path != local_path:
                        delete_document(previous_path)
                write_document(local_path, body)

            entry = ManifestDocument(
                url=doc.url,
                local_path=relative_local_path,
                language=doc.language,
                sha256=sha,
                bytes=len(body),
                etag=result.headers.get("ETag") or result.headers.get("Etag"),
                last_modified=result.headers.get("Last-Modified"),
                fetched_at=now,
                checked_at=now,
                status="updated" if changed else "unchanged",
                error=None,
            )
            manifest.documents[doc.url] = entry

            stats = CrawlStats(
                discovered=stats.discovered,
                fetched=stats.fetched + 1,
                updated=stats.updated + (1 if changed else 0),
                unchanged=stats.unchanged + (0 if changed else 1),
                errors=stats.errors,
                pruned=stats.pruned,
            )

        if prune:
            to_remove = set(manifest.documents.keys()) - discovered_urls
            for url in to_remove:
                doc = manifest.documents.get(url)
                if doc is None:
                    continue
                if not dry_run:
                    delete_document(ensure_within_dir(output_root, output_root / doc.local_path))
                manifest.documents.pop(url, None)
            stats = CrawlStats(
                discovered=stats.discovered,
                fetched=stats.fetched,
                updated=stats.updated,
                unchanged=stats.unchanged,
                errors=stats.errors,
                pruned=stats.pruned + len(to_remove),
            )
            logger.info("prune done source=%s removed=%d", source.id, len(to_remove))

        manifest.generated_at = iso_now()
        manifest.entrypoints = [ep.__dict__ for ep in source.entrypoints]
        manifest.documents_order = [doc.url for doc in docs]
        if not dry_run:
            save_manifest(manifest_path, manifest)
            if source.id == "react-aria":
                self._write_rac_note(output_root)

        return stats

    def _build_document_refs(
        self,
        source: SourceConfig,
        discovered: List[DiscoveredDoc],
        url_language_hints: Dict[str, Set[str]],
    ) -> List[DocumentRef]:
        docs: List[DocumentRef] = []
        for doc in discovered:
            rel = doc.relative_path if doc.relative_path is not None else safe_relative_path(doc.url)
            hints = url_language_hints.get(doc.url, set())
            language_hint = next(iter(hints)) if len(hints) == 1 else None
            language, rel_without_lang = split_language_from_path(
                rel, source.default_language, language_hint=language_hint
            )
            rel_without_lang = collapse_index_path(rel_without_lang)
            if rel_without_lang.suffix == "":
                rel_without_lang = rel_without_lang.with_suffix(".html")

            local_path = source.output_dir / language / rel_without_lang
            docs.append(
                DocumentRef(
                    url=doc.url,
                    language=language,
                    relative_path=rel_without_lang,
                    local_path=local_path,
                )
            )
        return docs

    @staticmethod
    def _progress_interval(total: int) -> int:
        if total <= 0:
            return 1
        return min(50, max(1, total // 10))

    @staticmethod
    def _normalize_manifest_paths(manifest: Manifest, output_root: Path) -> None:
        for doc in manifest.documents.values():
            if not doc.local_path:
                continue
            local_path = Path(doc.local_path)
            if local_path.is_absolute():
                try:
                    doc.local_path = str(local_path.relative_to(output_root.resolve()).as_posix())
                except ValueError:
                    continue
                continue
            try:
                doc.local_path = str(local_path.relative_to(output_root).as_posix())
            except ValueError:
                doc.local_path = str(local_path.as_posix())

    @staticmethod
    def _write_rac_note(output_root: Path) -> None:
        note_path = output_root.parent / "rac.md"
        note = (
            "# React Aria Components note\n\n"
            "React Aria docs include React Aria Components (RAC) content.\n"
            "Source: https://react-aria.adobe.com/llms.txt\n\n"
            "Search for \"react-aria-components\" or \"React Aria Components\" to find RAC docs.\n"
        )
        note_path.write_text(note, encoding="utf-8")

File: crates/docs-crawler/src/crawler.rs
use crate::discovery::{DiscoveredDoc, discover_documents};
use crate::http_client::HttpClient;
use crate::models::{DocumentRef, Manifest, ManifestDocument, SourceConfig};
use crate::storage::{delete_document, load_manifest, save_manifest, write_document};
use crate::util::{
    CrawlStats, collapse_index_path, ensure_within_dir, iso_now, normalize_url, path_to_posix,
    safe_relative_path, sha256_bytes, split_language_from_path,
};
use log::{info, warn};
use rayon::prelude::*;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicUsize, Ordering};

pub struct Crawler {
    allowed_extensions: Vec<String>,
    user_agent: String,
    timeout_seconds: u64,
    max_workers: usize,
}

impl Crawler {
    pub fn new(
        allowed_extensions: &[String],
        user_agent: &str,
        timeout_seconds: u64,
        max_workers: usize,
    ) -> Self {
        Self {
            allowed_extensions: allowed_extensions.to_vec(),
            user_agent: user_agent.to_string(),
            timeout_seconds,
            max_workers,
        }
    }

    pub fn crawl_source(
        &self,
        source: &SourceConfig,
        prune: bool,
        dry_run: bool,
    ) -> Result<CrawlStats, String> {
        let output_root = &source.output_dir;
        let output_root_abs = absolute_path(output_root);
        info!(
            "crawl start source={} output={}",
            source.id,
            output_root.display()
        );
        if dry_run {
            info!("dry run enabled source={}", source.id);
        }
        if prune {
            info!("prune enabled source={}", source.id);
        }

        let allowed_extensions = source
            .allowed_extensions
            .as_ref()
            .unwrap_or(&self.allowed_extensions)
            .to_vec();

        let manifest_path = output_root.join(".manifest.json");
        let mut manifest = load_manifest(&manifest_path)?.unwrap_or_else(|| Manifest {
            version: 1,
            source_id: source.id.clone(),
            generated_at: iso_now(),
            entrypoints: Vec::new(),
            documents: HashMap::new(),
            documents_order: Vec::new(),
        });

        normalize_manifest_paths(&mut manifest, output_root, &output_root_abs);

        let http = HttpClient::new(self.timeout_seconds, &self.user_agent)
            .map_err(|err| format!("Failed to init HTTP client: {err}"))?;

        let mut discovered_urls = HashSet::new();
        let mut discovered_language_hints: HashMap<String, HashSet<String>> = HashMap::new();
        let mut discovered_in_order: Vec<DiscoveredDoc> = Vec::new();

        for entrypoint in &source.entrypoints {
            info!(
                "discover start source={} type={}",
                source.id, entrypoint.kind
            );
            let cache_dir = if entrypoint.kind == "github_tree" {
                Some(
                    entrypoint
                        .git_cache_dir
                        .as_deref()
                        .map(PathBuf::from)
                        .unwrap_or_else(|| source.output_dir.join(".git-cache")),
                )
            } else {
                None
            };

            let discovered_batch = discover_documents(
                entrypoint,
                &allowed_extensions,
                &http,
                &source.default_language,
                cache_dir.as_deref(),
            )
            .map_err(|err| err.to_string())?;

            let before = discovered_urls.len();
            for discovered in discovered_batch {
                let url = normalize_url(&discovered.url);
                if discovered_urls.insert(url.clone()) {
                    discovered_in_order.push(DiscoveredDoc {
                        url: url.clone(),
                        language_hint: discovered.language_hint.clone(),
                        relative_path: discovered.relative_path.clone(),
                    });
                }
                if let Some(hint) = discovered.language_hint {
                    discovered_language_hints
                        .entry(url)
                        .or_default()
                        .insert(hint);
                }
            }
            let added = discovered_urls.len().saturating_sub(before);
            info!(
                "discover done source={} added={} total={}",
                source.id,
                added,
                discovered_urls.len()
            );
        }

        let docs = build_document_refs(source, &discovered_in_order, &discovered_language_hints);
        let mut stats = CrawlStats {
            discovered: docs.len(),
            ..CrawlStats::default()
        };

        let total_docs = docs.len();
        info!(
            "fetch start source={} total={} workers={}",
            source.id, total_docs, self.max_workers
        );

        let mut tasks = Vec::with_capacity(docs.len());
        for doc in &docs {
            let mut if_none_match = None;
            let mut if_modified_since = None;
            if let Some(previous) = manifest.documents.get(&doc.url) {
                let previous_path = if previous.local_path.is_empty() {
                    None
                } else {
                    Some(PathBuf::from(&previous.local_path))
                };
                let previous_path = previous_path
                    .map(|path| ensure_within_dir(output_root, &output_root.join(path)))
                    .transpose()?;
                let target_path = ensure_within_dir(output_root, &doc.local_path)?;
                let cache_ok = previous_path
                    .as_ref()
                    .map(|path| path.exists())
                    .unwrap_or(false)
                    || target_path.exists();
                if cache_ok {
                    if_none_match = previous.etag.clone();
                    if_modified_since = previous.last_modified.clone();
                }
            }
            tasks.push(FetchTask {
                doc: doc.clone(),
                if_none_match,
                if_modified_since,
            });
        }

        let progress_every = progress_interval(total_docs);
        let completed = AtomicUsize::new(0);
        let pool = rayon::ThreadPoolBuilder::new()
            .num_threads(self.max_workers)
            .build()
            .map_err(|err| err.to_string())?;

        let results: Vec<(String, crate::http_client::FetchResult)> = pool.install(|| {
            tasks
                .par_iter()
                .map(|task| {
                    let result = http.fetch_bytes(
                        &task.doc.url,
                        None,
                        task.if_none_match.as_deref(),
                        task.if_modified_since.as_deref(),
                        None,
                    );
                    let done = completed.fetch_add(1, Ordering::SeqCst) + 1;
                    if done % progress_every == 0 || done == total_docs {
                        info!(
                            "fetch progress source={} {}/{}",
                            source.id, done, total_docs
                        );
                    }
                    (task.doc.url.clone(), result)
                })
                .collect()
        });

        let mut results_map: HashMap<String, crate::http_client::FetchResult> =
            results.into_iter().collect();

        let now = iso_now();
        for doc in &docs {
            let Some(mut result) = results_map.remove(&doc.url) else {
                continue;
            };

            let mut previous = manifest.documents.remove(&doc.url);
            if result.status == 304 {
                if let Some(mut prev) = previous.take() {
                    let target_path = ensure_within_dir(output_root, &doc.local_path)?;
                    let previous_path = if prev.local_path.is_empty() {
                        None
                    } else {
                        Some(PathBuf::from(&prev.local_path))
                    };
                    let previous_path = previous_path
                        .map(|path| ensure_within_dir(output_root, &output_root.join(path)))
                        .transpose()?;

                    if !target_path.exists()
                        && previous_path
                            .as_ref()
                            .map(|path| path.exists())
                            .unwrap_or(false)
                    {
                        if !dry_run {
                            if let Some(parent) = target_path.parent() {
                                fs::create_dir_all(parent).map_err(|err| {
                                    format!("Failed to create dir {}: {err}", parent.display())
                                })?;
                            }
                            if let Some(previous_path) = previous_path.as_ref() {
                                fs::rename(previous_path, &target_path).map_err(|err| {
                                    format!(
                                        "Failed to move {} -> {}: {err}",
                                        previous_path.display(),
                                        target_path.display()
                                    )
                                })?;
                            }
                        }
                        let relative_target = target_path
                            .strip_prefix(&output_root_abs)
                            .unwrap_or(&target_path);
                        prev.local_path = path_to_posix(relative_target);
                        prev.checked_at = Some(now.clone());
                        prev.status = Some("moved".to_string());
                        manifest.documents.insert(doc.url.clone(), prev);
                        stats.updated += 1;
                        continue;
                    }

                    if !target_path.exists() {
                        let refreshed = http.fetch_bytes(&doc.url, None, None, None, None);
                        result = refreshed;
                        prev.checked_at = Some(now.clone());
                        prev.status = Some("refetched".to_string());
                        previous = Some(prev);
                    } else {
                        prev.checked_at = Some(now.clone());
                        prev.status = Some("not_modified".to_string());
                        manifest.documents.insert(doc.url.clone(), prev);
                        stats.unchanged += 1;
                        continue;
                    }
                }
            }

            let local_path = ensure_within_dir(output_root, &doc.local_path)?;
            let relative_local = local_path
                .strip_prefix(&output_root_abs)
                .unwrap_or(&local_path);
            let relative_local_path = path_to_posix(relative_local);

            if result.status != 200 || result.body.is_none() {
                warn!(
                    "fetch error source={} url={} status={} error={}",
                    source.id,
                    doc.url,
                    result.status,
                    result.error.clone().unwrap_or_default()
                );
                let mut entry = previous.unwrap_or_else(|| ManifestDocument {
                    url: doc.url.clone(),
                    local_path: relative_local_path.clone(),
                    language: doc.language.clone(),
                    sha256: String::new(),
                    bytes: 0,
                    ..ManifestDocument::default()
                });
                entry.checked_at = Some(now.clone());
                entry.status = Some("error".to_string());
                entry.error = Some(
                    result
                        .error
                        .unwrap_or_else(|| format!("status={}", result.status)),
                );
                manifest.documents.insert(doc.url.clone(), entry);
                stats.errors += 1;
                continue;
            }

            let body = result.body.take().unwrap_or_default();
            let sha = sha256_bytes(&body);

            let path_changed = previous
                .as_ref()
                .map(|doc| !doc.local_path.is_empty() && doc.local_path != relative_local_path)
                .unwrap_or(false);
            let changed = path_changed
                || previous
                    .as_ref()
                    .map(|doc| doc.sha256 != sha)
                    .unwrap_or(true)
                || !local_path.exists();

            if changed && !dry_run {
                if path_changed {
                    if let Some(prev) = previous.as_ref() {
                        if !prev.local_path.is_empty() {
                            let previous_path = ensure_within_dir(
                                output_root,
                                &output_root.join(PathBuf::from(&prev.local_path)),
                            )?;
                            if previous_path.exists() && previous_path != local_path {
                                delete_document(&previous_path)?;
                            }
                        }
                    }
                }
                write_document(&local_path, &body)?;
            }

            let entry = ManifestDocument {
                url: doc.url.clone(),
                local_path: relative_local_path,
                language: doc.language.clone(),
                sha256: sha,
                bytes: body.len(),
                etag: header_value(&result.headers, "etag"),
                last_modified: header_value(&result.headers, "last-modified"),
                fetched_at: Some(now.clone()),
                checked_at: Some(now.clone()),
                status: Some(if changed {
                    "updated".to_string()
                } else {
                    "unchanged".to_string()
                }),
                error: None,
            };
            manifest.documents.insert(doc.url.clone(), entry);

            stats.fetched += 1;
            if changed {
                stats.updated += 1;
            } else {
                stats.unchanged += 1;
            }
        }

        if prune {
            let to_remove: Vec<String> = manifest
                .documents
                .keys()
                .filter(|url| !discovered_urls.contains(*url))
                .cloned()
                .collect();
            for url in &to_remove {
                if let Some(doc) = manifest.documents.get(url) {
                    if !dry_run {
                        let path = ensure_within_dir(
                            output_root,
                            &output_root.join(PathBuf::from(&doc.local_path)),
                        )?;
                        delete_document(&path)?;
                    }
                }
                manifest.documents.remove(url);
            }
            stats.pruned += to_remove.len();
            info!(
                "prune done source={} removed={}",
                source.id,
                to_remove.len()
            );
        }

        manifest.generated_at = iso_now();
        manifest.entrypoints = source
            .entrypoints
            .iter()
            .map(|ep| serde_json::to_value(ep).unwrap_or(serde_json::Value::Null))
            .collect();
        manifest.documents_order = docs.iter().map(|doc| doc.url.clone()).collect();

        if !dry_run {
            save_manifest(&manifest_path, &manifest)?;
            if source.id == "react-aria" {
                write_rac_note(output_root)?;
            }
        }

        Ok(stats)
    }
}

#[derive(Clone)]
struct FetchTask {
    doc: DocumentRef,
    if_none_match: Option<String>,
    if_modified_since: Option<String>,
}

fn build_document_refs(
    source: &SourceConfig,
    discovered: &[DiscoveredDoc],
    url_language_hints: &HashMap<String, HashSet<String>>,
) -> Vec<DocumentRef> {
    let mut docs = Vec::with_capacity(discovered.len());
    for doc in discovered {
        let rel = doc
            .relative_path
            .clone()
            .unwrap_or_else(|| safe_relative_path(&doc.url));
        let hints = url_language_hints
            .get(&doc.url)
            .cloned()
            .unwrap_or_default();
        let language_hint = if hints.len() == 1 {
            hints.iter().next().cloned()
        } else {
            None
        };
        let (language, mut rel_without_lang) =
            split_language_from_path(&rel, &source.default_language, language_hint.as_deref());
        rel_without_lang = collapse_index_path(&rel_without_lang);
        if rel_without_lang.extension().is_none() {
            rel_without_lang.set_extension("html");
        }

        let local_path = source
            .output_dir
            .join(&language)
            .join(rel_without_lang.clone());
        docs.push(DocumentRef {
            url: doc.url.clone(),
            language,
            local_path,
        });
    }
    docs
}

fn progress_interval(total: usize) -> usize {
    if total == 0 {
        1
    } else {
        (total / 10).clamp(1, 50)
    }
}

fn normalize_manifest_paths(manifest: &mut Manifest, output_root: &Path, output_root_abs: &Path) {
    for doc in manifest.documents.values_mut() {
        if doc.local_path.is_empty() {
            continue;
        }
        let local_path = PathBuf::from(&doc.local_path);
        if local_path.is_absolute() {
            if let Ok(rel) = local_path.strip_prefix(output_root_abs) {
                doc.local_path = path_to_posix(rel);
            }
            continue;
        }
        if let Ok(rel) = local_path.strip_prefix(output_root) {
            doc.local_path = path_to_posix(rel);
        } else {
            doc.local_path = path_to_posix(&local_path);
        }
    }
}

fn header_value(headers: &HashMap<String, String>, key: &str) -> Option<String> {
    let lower = key.to_lowercase();
    if let Some(value) = headers.get(&lower) {
        return Some(value.clone());
    }
    headers
        .iter()
        .find(|(k, _)| k.eq_ignore_ascii_case(&lower))
        .map(|(_, v)| v.clone())
}

fn write_rac_note(output_root: &Path) -> Result<(), String> {
    let note_path = output_root
        .parent()
        .unwrap_or_else(|| Path::new("."))
        .join("rac.md");
    let note = concat!(
        "# React Aria Components note\n\n",
        "React Aria docs include React Aria Components (RAC) content.\n",
        "Source: https://react-aria.adobe.com/llms.txt\n\n",
        "Search for \"react-aria-components\" or \"React Aria Components\" to find RAC docs.\n"
    );
    fs::write(&note_path, note)
        .map_err(|err| format!("Failed to write {}: {err}", note_path.display()))?;
    Ok(())
}

fn absolute_path(path: &Path) -> PathBuf {
    if path.is_absolute() {
        path.to_path_buf()
    } else {
        std::env::current_dir()
            .unwrap_or_else(|_| PathBuf::from("."))
            .join(path)
    }
}
